### ISSUE TITLE
Istio-Proxy returning UH / no healthy upstream at startup for STRICT_DNS clusters 

### ISSUE THREAD

**Issue by nirvanagit** [istio/istio#48692]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

We have around 3-4k STRICT_DNS clusters (via ServiceEntries), and observed during startup istio-proxy returns a bunch of UH / no healthy upstream, which subsequently subside.

Our assumption is - Given the large amount of STRICT_DNS clusters, istio-proxy is taking time to complete the DNS resolution, and until then it doesn't show the endpoints for such clusters (checked with istioctl proxy-config endpoints command), and hence returns a UH / no healthy upstream.

Istio-proxy should not call itself ready if it is not able to handle calls to STRICT_DNS clusters, as it has not completed DNS resolution for them. 

Questions:
1. Can istio-proxy ensure all DNS resolutions are complete before taking any traffic?
2. Can it do runtime DNS resolution if it was not already done, and avoid UH / no healthy upstream errors? 

### Version

```bash
$ istioctl version
client version: 1.17.4
control plane version: 1.17-dca7c
data plane version: 1.17-dca7c (11 proxies)
```

```bash
╰─$ k version                                                                                                                                                                       11:03AM
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:47:38Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"25+", GitVersion:"v1.25.16-eks-8cb36c9", GitCommit:"3a3ea80e673d7867f47bdfbccd4ece7cb5f4a83a", GitTreeState:"clean", BuildDate:"2023-11-22T21:53:22Z", GoVersion:"go1.20.10", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.27) and server (1.25) exceeds the supported minor version skew of +/-1
```

**Comment by nirvanagit** [comment]
Thanks for the prompt response! 
Is istio proxy not marked READY when cluster is in warming state?

**Comment by howardjohn** [comment]
Yep, or it should behave that way anyways

**Comment by nirvanagit** [comment]
But we observed otherwise. We saw a proxy pod come up and return UH / no healthy upstream for STRICT_DNS endpoints. This error subsequently subsided after few seconds.

**Comment by nirvanagit** [comment]
Related issue on envoyproxy https://github.com/envoyproxy/envoy/issues/14670

**Comment by nirvanagit** [comment]
@howardjohn - could it be, that envoy did wait for the dns resolution of all STRICT_DNS clusters, but for a few it got a failure, but envoy still went ahead and marked itself as ready?

**Comment by howardjohn** [comment]
Seems plausible. Some debug logs may give more insight

**Comment by nirvanagit** [comment]
Got it. And when this happens, if the request comes for the failed DNS name, it would result in UH / no healthy upstream?

**Comment by nirvanagit** [comment]
Can we force envoy to do a DNS resolution during runtime, if the DNS resolution had failed earlier?

**Comment by howardjohn** [comment]
Envoy should retry the dns request if it fails (and if it succeeds)

**Comment by nirvanagit** [comment]
At runtime?

**Comment by nirvanagit** [comment]
> Envoy should retry the dns request if it fails (and if it succeeds)

Are you saying this based on the dns refresh rate configuration?

**Comment by howardjohn** [comment]
yes (and respect_dns_ttl)

**Comment by nirvanagit** [comment]
Got it, but can it be forced to do it at runtime if it was not already resolved, (in other words, it would be like adding a retry on no healthy upstream / UH errors)

**Comment by howardjohn** [comment]
No, I think its just based on the interval not "on demand" like this.

**Comment by nirvanagit** [comment]
I reproduced the issue, with additional monitoring in place, and this is the observation

1. istio proxy made all DNS queries successfully. Validated this by using `curl -s localhost:15000/stats/prometheus | grep envoy_dns_cares` metrics, and also validated that DNS resolution was successful by looking at node-local-dns logs
2. After 1 finished, still observed UH.

Could it be that envoy has received the resolution from the DNS in time, but it has not loaded them in its memory because of the large number of STRICT_DNS clusters?

**Comment by nirvanagit** [comment]
> update_failure

@yuval-k - don't have the update_failure metric right now. Will get it in the next simulation.
Regarding `/clusters` --> I don't see the expected count after envoy becomes ready. It eventually gets all the clusters (by clusters here I specifically mean, clusters with IP address).

**Comment by nirvanagit** [comment]
@yuval-k - will envoy mark itself ready of there was a cluster update failure?

**Comment by nirvanagit** [comment]
Latest observation:
1. We observed no errors reported by envoy for dns resolution
2. No errors reported by dns server
3. Envoy marks itself ready before making all the STRICT_DNS resolution calls to the DNS server

Is there some edge case with wait_for_warm_on_init ([link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto)), where it doesn't wait for making all the DNS queries? 

If we still think that envoy does make all the dns queries before calling itself ready, is there any data point I can use which shows that envoy made all the calls before marking itself ready?

**Comment by yuval-k** [comment]
>  will envoy mark itself ready of there was a cluster update failure?

Yes, if a DNS resolution has errored, envoy will consider that cluster warm.

When envoy starts a resolution, it will increment the `update_attempt` state  and when it finishes it will increment the `update_success` or `update_failure` stats accordingly.
When envoy got an update from DNS, but that did not change the hosts in the cluster, `update_no_rebuild` will be incremented.

When envoy is in the bad state, what are the values of these stats (note that theses stats are per-cluster)?

> If we still think that envoy does make all the dns queries before calling itself ready, is there any data point I can use which shows that envoy made all the calls before marking itself ready?

If the `update_success` for all the DNS clusters is not zero while the `/ready` admin page endpoint doesn't return 200 / "LIVE".

> Any workarounds for ensuring envoy doesn't mark itself ready before ensuring all DNS resolutions are done?

It's hard to provide a workaround without fully understanding the problem. There are other scenarios that can cause a failure here. for example:
- DNS resolution success but returns an empty list for one of the hosts in the cluster
- Envoy doesn't mark itself as ready, but from some reason LB thinks that it is

> How can I check that wait_for_warm_on_init is set?

you should be able to see if wait_for_warm_on_init is set in envoy's `/config_dump` in each of the clusters. If you don't see it, it defaults to `true`.

Also, it's worth getting the `/server_info` or `/ready` from the admin page to double check that envoy sees it self as LIVE.

**Comment by nirvanagit** [comment]
Enabled dns:debug on envoy and found that there are a bunch of failed dns resolutions getting logged as:
```bash
2024-01-11T19:15:54.069747Z	debug	envoy dns external/envoy/source/extensions/network/dns_resolver/cares/dns_impl.cc:153	dns resolution for internal-a4123445abab34fdf8b316aeff11bc2e-712665588.us-west-2.elb.amazonaws.com failed with c-ares status 12	thread=24
```

Even though, DNS Metrics emitted by envoy do not indicate any failures:
```bash
# TYPE envoy_dns_cares_get_addr_failure counter
envoy_dns_cares_get_addr_failure{} 0
# TYPE envoy_dns_cares_not_found counter
envoy_dns_cares_not_found{} 1629
# TYPE envoy_dns_cares_resolve_total counter
envoy_dns_cares_resolve_total{} 20927
# TYPE envoy_dns_cares_pending_resolutions gauge
envoy_dns_cares_pending_resolutions{} 504

```

**Comment by nirvanagit** [comment]
In the node local DNS logs, 63 unique DNS resolutions occurred before Envoy declared readiness. However, envoy’s DNS logs revealed 73 queries (both successful and failed) before the same readiness marker. (FYI: We expect 73 unique dns names)
Further analysis indicates that failed resolutions were absent in node-local-dns logs, potentially due to limited logging for certain failure types. Envoy’s DNS logs suggest some failures before Envoy marked itself ready. While most failed resolutions persisted after readiness, one DNS record succeeded post-Envoy readiness. _This proves that Envoy does all the DNS resolutions, some of which might fail, and that doesn’t stop Envoy from marking itself ready._

Question:
1. Is it possible to add retries in envoy for DNS resolution, and lets mark itself READY only after all the retries have been completed (failed / success).
2. Is it possible to prioritize DNS resolution for certain records? 
3. What metric could I use to validate this data, because the DNS metrics I posted above, do not indicate such failures.

**Comment by yuval-k** [comment]
you can see the ares error codes here:

https://github.com/c-ares/c-ares/blob/2a6a420cb6cdcc4f00b3114df9b4a0b02d165ddf/include/ares.h#L122

```
/* Server error codes (ARES_ENODATA indicates no relevant answer) */
#define ARES_ENODATA            1
...
#define ARES_ETIMEOUT           12
```

you should be able to see the timeouts (error 12) in `envoy_dns_cares_timeouts` and no data (error 1) in  `envoy_dns_cares_not_found` .

**Comment by yuval-k** [comment]
Regarding questions 1 & 2 - I don't believe this is possible in envoy today.

**Comment by nirvanagit** [comment]
More details:
1. Looked into DNS constraints in AWS EC2 instances. And the limit has not been breached
2. Tested with 3k service entries using STRICT_DNS, and that reduced the number of instances which encountered this issue. And the errors also reduced significantly.
3. Looked into https://github.com/envoyproxy/envoy/issues/7965, which mentions how c-ares does DNS resolution when u sing UDP. It opens one socket, and does all the DNS resolutions using that. 
    So it could be that the socket is overwhelmed. Tried TCP, and saw the various errors disappear, except timeouts. 
    
    So far it appears DNS of UDP is an issue, but switching over to TCP is resulting in a spike in timeouts (and no other errors, like status 1, 11, 16). 
    Question: How can I increase the TCP timeout for DNS resolutions?

**Comment by nirvanagit** [comment]
@yuval-k @howardjohn - is this the correct way to enable DNS over TCP?
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: dns-over-tcp
  namespace: test
spec:
  workloadSelector:
    labels:
      app: test
  configPatches:
  - applyTo: CLUSTER
    patch:
      operation: MERGE
      value:
        use_tcp_for_dns_lookups: true
```

I added this configuration, but when I do `netstat -tunapl | grep <DNS_IP>` I see it is still using udp.
```bash
udp        0      0 10.193.200.3:32993      172.20.0.10:53          ESTABLISHED 26/envoy
```

**Comment by nirvanagit** [comment]
Tried this:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: dns-over-tcp
  namespace: test
spec:
  workloadSelector:
    labels:
      app: test
  configPatches:
  - applyTo: CLUSTER 
    patch:
      operation: MERGE
      value:
        #use_tcp_for_dns_lookups: true
        dns_resolution_config:
          #typed_config:
          #"@type": "type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig"
          dns_resolver_options:
            use_tcp_for_dns_lookups: true 
```

**Comment by yuval-k** [comment]
> 
> ```shell
> udp        0      0 10.193.200.3:32993      172.20.0.10:53          ESTABLISHED 26/envoy
> ```

as this config applies to dynamic clusters coming from istio, envoy might still use UDP for static clusters (for example the XDS cluster - istiod). Do you not see tcp used at all?

Regarding the second error, envoy rejected your CDS config. This probably means that the patch did not apply as expected.. Usually there's more information in the envoy log or the config_dump as to why the cluster is invalid. do you see anything in the log (probably above that line) to indicate the specific error?

**Comment by nirvanagit** [comment]
> > ```shell
> > udp        0      0 10.193.200.3:32993      172.20.0.10:53          ESTABLISHED 26/envoy
> > ```
> 
> as this config applies to dynamic clusters coming from istio, envoy might still use UDP for static clusters (for example the XDS cluster - istiod). Do you not see tcp used at all?
> 
> Regarding the second error, envoy rejected your CDS config. This probably means that the patch did not apply as expected.. Usually there's more information in the envoy log or the config_dump as to why the cluster is invalid. do you see anything in the log (probably above that line) to indicate the specific error?

Don't see anything go out to the dns server at all with the first envoyfilter configured.

**Comment by asayah** [comment]
can you try this, this worked for me: 
```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: dns-over-tcp
  namespace: istio-system
spec:
  configPatches:
  - applyTo: CLUSTER
    patch:
      operation: MERGE
      value:
        typed_dns_resolver_config:
          name: envoy.network.dns_resolver.cares
          typedConfig:
            '@type': type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig
            dns_resolver_options:
              use_tcp_for_dns_lookups: true
```

result in
```
  type: STRICT_DNS
  typedDnsResolverConfig:
    name: envoy.network.dns_resolver.cares
    typedConfig:
      '@type': type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig
      dnsResolverOptions:
        useTcpForDnsLookups: true
```

**Comment by nirvanagit** [comment]
Got this working:
```yaml
  configPatches:
  - applyTo: CLUSTER 
    patch:
      operation: MERGE
      value:
        #use_tcp_for_dns_lookups: true
        #respect_dns_ttl: false 
        dns_resolution_config:
          #typed_config:
          #"@type": "type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig"
          resolvers:
          - socket_address: 
              address: "172.20.0.10"
              port_value: "53"
          dns_resolver_options:
            use_tcp_for_dns_lookups: true 
```

**Comment by nirvanagit** [comment]
> can you try this, this worked for me:
> 
> ```
> apiVersion: networking.istio.io/v1alpha3
> kind: EnvoyFilter
> metadata:
>   name: dns-over-tcp
>   namespace: istio-system
> spec:
>   configPatches:
>   - applyTo: CLUSTER
>     patch:
>       operation: MERGE
>       value:
>         typed_dns_resolver_config:
>           name: envoy.network.dns_resolver.cares
>           typedConfig:
>             '@type': type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig
>             dns_resolver_options:
>               use_tcp_for_dns_lookups: true
> ```
> 
> result in
> 
> ```
>   type: STRICT_DNS
>   typedDnsResolverConfig:
>     name: envoy.network.dns_resolver.cares
>     typedConfig:
>       '@type': type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig
>       dnsResolverOptions:
>         useTcpForDnsLookups: true
> ```
> 
> I didn't check if it switch the resolution to use TCP though

Thanks @asayah - tried that one, but that doesn't change the communication protocol.

**Comment by nirvanagit** [comment]
> Got this working:
> 
> ```yaml
>   configPatches:
>   - applyTo: CLUSTER 
>     patch:
>       operation: MERGE
>       value:
>         #use_tcp_for_dns_lookups: true
>         #respect_dns_ttl: false 
>         dns_resolution_config:
>           #typed_config:
>           #"@type": "type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig"
>           resolvers:
>           - socket_address: 
>               address: "172.20.0.10"
>               port_value: "53"
>           dns_resolver_options:
>             use_tcp_for_dns_lookups: true 
> ```

But this one worked. Evaluating if switching to TCP is helping.

**Comment by nirvanagit** [comment]
I see envoy is creating new TCP connections for DNS resolutions. And it is creating multiple such connections, (whereas UDP creates only one socket).

Is there a way we could have long running TCP connections for DNS resolutions? Didn't find any configuration in envoy for DNS resolution specifically.

**Comment by yuval-k** [comment]
The TCP DNS connections are created by the cares library, so envoy doesn't have much control of their behaviour

**Comment by nirvanagit** [comment]
Is it possible to increase the timeout for DNS resolution? I see there is an option in the c-ares library `ARES_OPT_TIMEOUT`, but don't see an option to override that in [envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/resolver.proto#envoy-v3-api-msg-config-core-v3-dnsresolveroptions)

**Comment by zirain** [comment]
> Is it possible to increase the timeout for DNS resolution? I see there is an option in the c-ares library `ARES_OPT_TIMEOUT`, but don't see an option to override that in [envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/resolver.proto#envoy-v3-api-msg-config-core-v3-dnsresolveroptions)

can you raise this up in envoy first?

**Comment by nirvanagit** [comment]
> > Is it possible to increase the timeout for DNS resolution? I see there is an option in the c-ares library `ARES_OPT_TIMEOUT`, but don't see an option to override that in [envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/resolver.proto#envoy-v3-api-msg-config-core-v3-dnsresolveroptions)
> 
> can you raise this up in envoy first?

https://github.com/envoyproxy/envoy/issues/31992
